python/mlx/nn/losses.py

```diff
@@ -15,24 +15,57 @@ def decorator(klass):
 
 
 def cross_entropy(
-    logits: mx.array, targets: mx.array, axis: int = -1, reduction: str = "none"
+    logits: mx.array,
+    targets: mx.array,
+    weights: mx.array = None,
+    axis: int = -1,
+    label_smoothing: float = 0.0,
+    reduction: str = "none",
 ) -> mx.array:
     """
     Computes the cross entropy loss between logits and targets.
 
     Args:
         logits (mx.array): The predicted logits.
-        targets (mx.array): The target values.
+        targets (mx.array): The target values, as class indices.
+        weights (mx.array, optional): Weights for each target. Default: ``None``.
         axis (int, optional): The axis over which to compute softmax. Default: ``-1``.
+        label_smoothing (float, optional): Label smoothing factor. Default: ``0``.
         reduction (str, optional): Specifies the reduction to apply to the output:
-          ``'none'`` | ``'mean'`` | ``'sum'``. Default: ``'none'``.
+            ``'none'`` | ``'mean'`` | ``'sum'``. Default: ``'none'``.
 
     Returns:
         mx.array: The computed cross entropy loss.
     """
+    assert (
+        0 <= label_smoothing < 1
+    ), "label smoothing must be between 0 and 1 (exclusive)"
+    assert (
+        logits.shape[0] == targets.shape[0]
+    ), "The shape of logits and targets must match along axis 0."
+
     score = mx.take_along_axis(logits, targets[..., None], axis).squeeze(-1)
-    loss = mx.logsumexp(logits, axis=axis) - score
+    logsumexp_logits = mx.logsumexp(logits, axis=axis)
+    if label_smoothing > 0:
+        # Adjust the true class score with label smoothing
+        adjusted_score = (1 - label_smoothing) * score
+
+        # Calculate the mean logit across the classes for smoothed loss
+        mean_logits = logits.mean(axis=axis)
+        smoothed_loss = -mean_logits * label_smoothing
+
+        # Combine the adjusted score and smoothed loss with the logsumexp logits
+        loss = logsumexp_logits - adjusted_score + smoothed_loss
```

Comment:
Are you basing this off of a reference implementation? Just curious as I see label smoothing defined as `(1-l)` for the ground truth and `l / (N -1)` for the remaining labels. where `l` is the label smoothing parameter. 

But here the label smoothing parameter gives you `1 - l + l/N` for the ground truth so it's a slightly different parameterization. I like the simplicity of implementation, but I'm just wondering what's standard and if we should follow that?

Reply:
There are many implementations out there I believe. I referred to some of them [here](https://stackoverflow.com/questions/55681502/label-smoothing-in-pytorch), though it's not an exact replica of any particular implementation. I kinda adapted them to optimize computation by reusing `logsumexp` and `score`.

From my understanding, label smoothing reduces the confidence of the correct class $y_{c} = 1$ by $l$ and redistributes $l$ evenly among all classes, which includes the correct class too. Hence, the smoothed label for the correct class becomes $1 - l + \frac{l}{N}$, and each incorrect class gets $\frac{l}{N}$, where $N$ is the total number of classes.

The standard cross-entropy without label smoothing is given as:

$$
\begin{aligned}
CE(\textbf{x},\textbf{y}) &= - \sum_{k=1}^{N} y_k \log(p(x_k)) \\
&= - \sum_{k=1}^{N} y_k \log\left(\frac{\exp(x_k)}{\sum_{i} \exp(x_i)}\right) \\
&= - \sum_{k=1}^{N} y_k \left( x_k - \log\left( \sum_{i=1}^{N} \exp(x_i) \right) \right) \\
&=  \log\left( \sum_{i=1}^{N} \exp(x_i) \right) - \sum_{k=1}^{N} y_k x_k 
\end{aligned}
$$

Including label smoothing, the term $\sum_{k} y_k x_k$ is recalculated as:

$$
\begin{aligned}
\sum_{k=1}^{N} y_k x_k
&= \left( (1 - l) + \frac{l}{N} \right) x_c + \sum_{k \neq c} \frac{l}{N} x_k \\
&= (1 - l) x_c + \frac{l}{N} \sum_{k=1}^{N} x_k
\end{aligned}
$$

The contributions from the modified labels sum to 1:

- For the correct class $c$, the contribution is $(1 - l) + \frac{l}{N}$.
- For each of the $N - 1$ incorrect classes, the contribution is $\frac{l}{N}$. Summing this over all $N - 1$ classes gives $(N - 1) \times \frac{l}{N}$.

The total sum is then:

$$
(1 - l) + \frac{l}{N} + (N - 1) \times \frac{l}{N} = 1
$$

This shows that after label smoothing, the sum of the weighted labels remains 1, maintaining the property of a probability distribution.

I hope that answers your question!

Reply:
What you are saying makes perfect sense, and thanks for the details. I had no doubt it was a valid distribution you are computing, my only point is that an alternative parameterization is to have the ground truth get a weight of $(1-l)$ instead of $(1-l) + /frac{1}{N}$. In that case the other labels get a weight of $\frac{l}{N-1}$.

That said, what you have is definitely cleaner/simpler to implement so I'm in favor of keeping it the way it is.

Reply:
Ah, I see where you're coming from. Sorry, I misunderstood. You're right that there's an alternative that uses $1-l$ for the ground truth and $\frac{l}{N-1}$ for the other incorrect labels. I'm not really sure which is standard as there are many implementations that use either one. In this regard, I'm following [Nvidia](https://github.com/NVIDIA/DeepLearningExamples/blob/8d8b21a933fff3defb692e0527fca15532da5dc6/PyTorch/Classification/ConvNets/image_classification/smoothing.py#L18) in using $1-l+\frac{l}{N}$ for ground truth.